### ISSUE TITLE
Multithreaded execution

### ISSUE THREAD

**Issue by barrucadu** [barrucadu/dejafu#126]
dejafu executes a program-under-test all in a single thread.  This simplifies some things, but means that testing things using bound threads is impossible.  Bound threads are necessary for FFI calls which use thread-local state, so it's pretty unfortunate that these are out of reach.

On the other hand, dejafu could create actual threads, and schedule them by using `MVar`s or something.  Each worker thread would look something like:

```haskell
work in_ out_ = go where
  go act = do
    ctx <- takeMVar in_
    (act', trace, ctx') <- doAct act
    putMVar out_ (trace, ctx')
    go act'
```

Then the main loop would use the scheduler to pick a thread to run, single-step it by putting once to its input `MVar`, and update the global state and trace by reading from the output `MVar`.

This would let us test bound threads: just fork one for the worker!  *However*, it would add considerable overhead *and* be a breaking API change.  Overhead because suddenly we have lots of extra threads and `MVar`s, breaking API change because `ConcT` would now need to take a `MonadConc`.  So it would look something like this:

```haskell
dejafu :: (MonadConc m, MonadIO m) => DejaFu m a -> (String, Predicate a) -> m Bool
runConcurrent :: MonadConc m => Scheduler s -> MemType -> s -> DejaFu m a -> m (Either Failure a, s, Trace)
runSCT :: MonadConc m => Way -> MemType -> DejaFu m a -> m [(Either Failure a, Trace)]
```

A `ConcST` would no longer be possible, as `ST` can't create bound threads (or any threads).  So we'd gain the ability to test something new at the cost of purity.  On the other hand, supporting both `IO` and `ST` currently leads to a lot of duplication, which would simply vanish after this change.

I don't think any change to Test.DejaFu.SCT would be necessary (other than changing type signatures).  In fact, the logic changes should be entirely restricted to Test.DejaFu.Conc.Internal.

**Comment by barrucadu** [comment]
The overhead could probably be avoided in the common case by only forking threads if necessary (ie, they're bound), and keeping the current single-threaded logic for unbound threads.  But that could complicate the implementation.

Worth trying out though.

**Comment by barrucadu** [comment]
Implemented in bfe04d2 (the big API change) and 8381138 (the actual bound threads implementation).

**Comment by barrucadu** [comment]
Implemented in `next-supermajor` branch.
